src/providers/vmware/amd64.rs

```diff
@@ -39,8 +63,155 @@ impl VmwareProvider {
+            // YAML we don't need to try serde_json::from_str here.
+            let netplan_config_unfiltered: serde_json::Value = serde_yaml::from_str(metadata)?;
+            // Only the "network" key is allowed to be present.
+            // We use the json! macro but this is only about creating a serde value::Value,
+            // even though its name sounds like it would create JSON.
+            let netplan_config_filtered = json!({ "network": netplan_config_unfiltered.get("network").ok_or(anyhow!("no 'network' key found"))?});
+            let netplan_config = serde_yaml::to_string(&netplan_config_filtered)?;
+            Ok(Some(netplan_config))
+        } else {
+            Ok(None)
+        }
+    }
+
+    #[cfg(test)]
+    pub fn new_from_metadata(metadata: String) -> Result<Self> {
+        Ok(Self {
+            guestinfo_net_kargs: None,
+            guestinfo_metadata: Some(metadata),
+        })
+    }
+}
+
+fn parse_metadata(
+    guestinfo_metadata_encoding: Option<String>,
+    guestinfo_metadata_raw: Option<String>,
+) -> Result<Option<String>> {
+    let guestinfo_metadata =
+        if let Some(guestinfo_metadata_encoding_val) = guestinfo_metadata_encoding {
+            match (
+                guestinfo_metadata_encoding_val.as_str(),
+                guestinfo_metadata_raw,
+            ) {
+                ("base64" | "b64", Some(guestinfo_metadata_raw_val)) => {
+                    let decoded =
+                        general_purpose::STANDARD.decode(guestinfo_metadata_raw_val.as_bytes())?;
+                    Some(String::from_utf8(decoded)?)
+                }
+                ("gzip+base64" | "gz+b64", Some(guestinfo_metadata_raw_val)) => {
+                    let decoded =
+                        general_purpose::STANDARD.decode(guestinfo_metadata_raw_val.as_bytes())?;
+                    let mut decompressor = Decoder::new(decoded.as_slice())?;
+                    let mut uncompressed = Vec::new();
+                    decompressor.read_to_end(&mut uncompressed)?;
+                    Some(String::from_utf8(uncompressed)?)
+                }
+                ("", guestinfo_metadata_raw) => guestinfo_metadata_raw,
+                (&_, _) => bail!("unknown guestinfo.metadata.encoding"),
+            }
+        } else {
+            guestinfo_metadata_raw
+        };
+    Ok(guestinfo_metadata)
+}
+
+#[test]
+fn test_netplan_json() {
+    let metadata = r#"{
+      "network": {
+        "version": 2,
+        "ethernets": {
+          "nics": {
+            "match": {
+              "name": "ens*"
+            },
+            "dhcp4": true
+          }
+        }
+      },
+      "ExcludeNonNetplanField": 0
+    }"#;
+    let provider = VmwareProvider::new_from_metadata(metadata.to_owned()).unwrap();
+    let netplan_config = provider.parse_netplan_config().unwrap().unwrap();
+    // Note: The test relies on the serde 'preserve_order' feature for equality
```

Comment:
It seems like that's the only reason we're adding `preserve_order`?

Can we avoid this by slightly tweaking the tests to just not have multiple subkeys in a key? We're not actually semantically parsing this anyway, and we're still testing the main thing, which is that we can parse JSON and YAML and exclude unknown top-level fields.

Reply:
Yes, done